src/pcp2openmetrics/pcp2openmetrics.py

```diff
@@ -0,0 +1,569 @@
+
+    def report(self, tstamp):
+        """ Report metric values """
+        if tstamp is not None:
+            tstamp = tstamp.strftime(self.timefmt)
+       
+        self.write_openmetrics(tstamp)
+
+    def write_header(self):
+        """ Write info header """
+        output = self.outfile if self.outfile else "stdout"
+        if self.context.type == PM_CONTEXT_ARCHIVE:
+            sys.stdout.write('{ "//": "Writing %d archived metrics to %s..." }\n{ "//": "(Ctrl-C to stop)" }\n' % (len(self.metrics), output))
+            return
+
+        sys.stdout.write('{ "//": "Waiting for %d metrics to be written to %s' % (len(self.metrics), output))
+        if self.runtime != -1:
+            sys.stdout.write(':" }\n{ "//": "%s samples(s) with %.1f sec interval ~ %d sec runtime." }\n' % (self.samples, float(self.interval), self.runtime))
+        elif self.samples:
+            duration = (self.samples - 1) * float(self.interval)
+            sys.stdout.write(':" }\n{ "//": "%s samples(s) with %.1f sec interval ~ %d sec runtime." }\n' % (self.samples, float(self.interval), duration))
+        else:
+            sys.stdout.write('..." }\n{ "//": "(Ctrl-C to stop)" }\n')
+
+    def write_openmetrics(self, timestamp):
+        """ Write results in openmetrics format """
+        if timestamp is None:
+            # Silent goodbye, close in finalize()
+            return
+
+        ts = self.context.datetime_to_secs(self.pmfg_ts(), PM_TIME_SEC)
+
+        if self.prev_ts is None:
+            self.prev_ts = ts
+
+        if not self.writer and not self.url:
+            if self.outfile is None:
+                self.writer = sys.stdout
+            else:
+                self.writer = open(self.outfile, 'wt')
+
+        def get_type_string(desc):
+            """ Get metric type as string """
+            if desc.contents.type == pmapi.c_api.PM_TYPE_32:
+                mtype = "32-bit signed"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_U32:
+                mtype = "32-bit unsigned"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_64:
+                mtype = "64-bit signed"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_U64:
+                mtype = "64-bit unsigned"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_FLOAT:
+                mtype = "32-bit float"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_DOUBLE:
+                mtype = "64-bit float"
+            elif desc.contents.type == pmapi.c_api.PM_TYPE_STRING:
+                mtype = "string"
+            else:
+                mtype = "unknown"
+            return mtype
+
+        def openmetrics_name(metric):
+            """ pcp.io metric name to openmetrics.io name conventions """
+            return metric.replace('.','_')
+        
+        def openmetrics_help(context, pmid):
+            """ convert pcp.io metric helptext to openmetrics.io HELP """
+            return context.pmLookupText(pmid[0]).decode()
```

Comment:
Hmm, pmLookupText involves a roundtrip to pmcd on every call (i.e. for every metric here).  We should move this out of the fetch loop - create a dictionary at the start which maps each metric to its help text, and do a lookup here instead of communicating with pmcd (we'll get the same answer every time for help text).